### ISSUE TITLE
Logistic Distribution

### ISSUE THREAD

**Issue by sameerkhurana10** [pytorch/pytorch#7857]
Could we have logistic distribution in PyTorch?

https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/distributions/python/ops/logistic.py

**Comment by fritzo** [comment]
We decided not to add this in #4833 since it's so easy to implement using `TransformedDistribution`:
```py
TransformedDistribution(Uniform(0, 1), [SigmoidTransform().inv, AffineTransform(a, b)])
```
Maybe it's worth adding this as an example in the `TransformedDistribution` docs?

**Comment by EtienneDesticourt** [comment]
I feel like there should be a warning/reminder along with this example that the distribution will sometimes return -inf when Uniform returns 0.

**Comment by ludkinm** [comment]
> We decided not to add this in #4833 since it's so easy to implement using `TransformedDistribution`:
> 
> ```python
> TransformedDistribution(Uniform(0, 1), [SigmoidTransform().inv, AffineTransform(a, b)])
> ```
> 
> Maybe it's worth adding this as an example in the `TransformedDistribution` docs?
Sorry to comment on an old and closed FR but I can't work out how to get this working in pyro.
I ran this snippet:
 ```python
def Logistic(a,b):
    return TransformedDistribution(Uniform(0, 1), [SigmoidTransform().inv, AffineTransform(a, b)])
```
which is ok, but later in a model i get the error `TransformedDistribution' object is not callable`. 
Do I need to make a `Logistic` class which inherits from `TransformedDistribution`?
Or is there some torch to pyro magic to do this?

**Comment by ludkinm** [comment]
In case anyone else arrives from google, I was importing from torch, not pyro which caused the errors:

Based on @fritzo's transformations
```python
from pyro.distributions import Uniform
from pyro.distributions.torch import TransformedDistribution
from pyro.distributions.transforms import SigmoidTransform, AffineTransform
def Logistic(a,b):
    return TransformedDistribution(Uniform(0, 1), [SigmoidTransform().inv, AffineTransform(a,b)])
```
Usage inside a model: `x = pyro.sample('x', Logistic(torch.ones(1), torch.ones(1)))`

**Comment by fritzo** [comment]
That's right, Pyro distributions also inherit from Pyro's [TorchDistributionMixin](http://docs.pyro.ai/en/stable/distributions.html#pyro.distributions.torch_distribution.TorchDistributionMixin) class which adds Pyro-specific functionality such as a `.__call__()` method and `.score_parts()` method for ELBO computations. When you import `TransformedDistribution` from Pyro you get a class that derives from both the `TransformedDistribution` in torch and the mixin class.

**Comment by KIC** [comment]
> We decided not to add this in #4833 since it's so easy to implement using `TransformedDistribution`:
> 
> ```python
> TransformedDistribution(Uniform(0, 1), [SigmoidTransform().inv, AffineTransform(a, b)])
> ```
> 
> Maybe it's worth adding this as an example in the `TransformedDistribution` docs?
sorry I don't get why you won't add something because it is easy. you might even just provide a constant like 
```
Logistik Distribution = TransformedDistribution(Uniform(0, 1), [SigmoidTransform().inv, AffineTransform(a, b)])
```

this way we as user can always be sure it is supported and works as expected. everything we handcrafted can contain a potential bug we have introduced into our model.

**Comment by fritzo** [comment]
@KIC point taken, feel free to add a simple `class Logistic(TransformedDistribution)` together with declared `.support`, docs, and tests in test/distributions/test_distrbutions.py.

**Comment by KIC** [comment]
Looks like after all it is not so easy. Check this example:

```python
from torch.distributions import Uniform, SigmoidTransform, AffineTransform, TransformedDistribution

def Logistic(loc, scale):
    return TransformedDistribution(Uniform(0, 1), [SigmoidTransform().inv, AffineTransform(loc, scale)])

cat = t.rand(10, 1)
dist = MixtureSameFamily(
    Categorical(probs=t.cat([1-cat, cat], dim=1)),
    Logistic(loc=t.randn(10,12), scale=t.ones(10, 2))
)

print(dist.sample())
```

sadly this errors out, and I can't get the shapes right ... 
~/.local/lib/python3.8/site-packages/torch/distributions/mixture_same_family.py in __init__(self, mixture_distribution, component_distribution, validate_args)
     75         # Check that the number of mixture component matches
     76         km = self._mixture_distribution.logits.shape[-1]
---> 77         kc = self._component_distribution.batch_shape[-1]
     78         if km is not None and kc is not None and km != kc:
     79             raise ValueError("`mixture_distribution component` ({0}) does not"

IndexError: tuple index out of range

**Comment by fritzo** [comment]
@KIC looks good to me, contribution welcome!  Actually I think it would be best to add a whole new subclass of TransformedDistribution, similar to [Pareto](https://github.com/pytorch/pytorch/blob/ef13341a8d1128a5fbd54716c89358cbd20c0a1d/torch/distributions/pareto.py#L30), that way it can provide metadata and properties like `.mean`:

<details>

```py
class Logistic(TransformedDistribution):
    """
    TODO add docs
    """
    arg_constraints = {"loc": constraints.real, "scale": constraints.positive}
    support = constraints.real
    has_rsample = True

    def __init__(self, loc, scale, validate_args=None):
        self.loc, self.scale = broadcast_all(loc, scale)
        base_dist = Uniform(self.loc.new_zeros(self.loc.shape),
                            self.scale.new_zeros(self.scale.shape))
        transforms = [SigmoidTransform().inv, AffineTransform(self.loc, self.scale)]
        super().__init__(base_dist, transforms, validate_args=validate_args)

    def expand(self, batch_shape):
        # TODO copy from another transformed distribution, e.g. Pareto

    @property
    def mean(self):
       return self.loc

    @property
    def variance(self):
        return (math.pi ** 2 / 3) * self.scale ** 2
```

</details>

**Comment by rfeinman** [comment]
FYI there is an equivalent distribution now in `LogitRelaxedBernoulli`. No transform needed.

```python
from torch.distributions.relaxed_bernoulli import LogitRelaxedBernoulli

def Logistic(loc, scale):
    return LogitRelaxedBernoulli(temperature=1/scale, logits=loc/scale)
```
